### DISCUSSION TITLE
[Berry]: Test for JSON key before accessing

### DISCUSSION THREAD

**Discussion by Slyke** [arendst/Tasmota#15160] [Q&A] [answered]
I have the following code:
```
      var sensorResult = json.load(tasmota.read_sensors())

      self.temperature = -275 # Default value to know that reading failed

      if sensorResult != nil
        if sensorResult['SHT3X'] != nil
          if sensorResult['SHT3X']['Temperature'] != nil
            self.temperature = sensorResult['SHT3X']['Temperature']
          end
        end
      end
```

How can I test if the `SHT3X` key exists, and if the `Temperature` key exists inside it? If I unplug the temperature module, this code errors with `key_error`.

I checked here: https://github.com/berry-lang/berry/wiki/Chapter-7#json-module But couldn't figure out an easy way to check for keys.

**Comment by sfromis** [comment, marked as answer]
What you ask about is not a JSON-related question, as the outcome of `json.load` is a standard map, with functions available no matter how it was created. 

To check for existence, you can use:
```
if sensorResult.contains('SHT3X')
  ...
end
```

You can also do things like:
```
var sht = sensorResult.find('SHT3X')
if sht != nil
  if sht.contains('Temperature')
    ...
  end
end
```

**Comment by s-hadinger** [comment]
You can even combine the `find` into a one-liner. The idea is that `find()` can specify a fallback value. If you want to cascade accessors, you can fallback to an empty map `[]`.

``` ruby
self.temperature = sensorResult.find('SHT3X', []).find('Temperature', -275) # Default value to know that reading failed
```

**Reply by sfromis** [reply]
What's the plan for getting such features into the docs? I was considering doing an edit, but with at least 3 places of seeing Berry docs, I did not go forward with that.

**Reply by s-hadinger** [reply]
Putting this in the Berry doc for `find()` could be a good start, but maybe in the Tasmota/Berry cookbook as well. Both are open to edits.

**Reply by sfromis** [reply]
What I meant was https://github.com/berry-lang/berry/wiki as referred to by Slyke, https://github.com/berry-lang/berry_doc and the PDF documents.

**Reply by s-hadinger** [reply]
Oh, I see. Please use https://github.com/berry-lang/berry/wiki
The others are outdated

**Reply by sfromis** [reply]
`find` and `contains` added to wiki.
